File: ruoyi-vpn-client/src/crypto/OfflineLoginIntegrity.cpp
#include "OfflineLoginIntegrity.h"

#include "HmacSha256.h"

#include <QByteArray>
#include <QJsonObject>

namespace vpn {

namespace {

QString fieldString(const QJsonObject &obj, const char *snakeKey, const char *camelKey = nullptr)
{
    if (obj.contains(snakeKey)) {
        return obj.value(snakeKey).toString();
    }
    if (camelKey && obj.contains(camelKey)) {
        return obj.value(camelKey).toString();
    }
    return QString();
}

int fieldInt(const QJsonObject &obj, const char *snakeKey, const char *camelKey)
{
    if (obj.contains(snakeKey)) {
        return obj.value(snakeKey).toInt();
    }
    if (camelKey && obj.contains(camelKey)) {
        return obj.value(camelKey).toInt();
    }
    return 0;
}

void appendLine(QString &out, const QString &key, const QString &value)
{
    out += key;
    out += QLatin1Char('=');
    out += value;
    out += QLatin1Char('\n');
}

void appendLine(QString &out, const QString &key, int value)
{
    appendLine(out, key, QString::number(value));
}

bool secureEquals(const QByteArray &a, const QByteArray &b)
{
    if (a.size() != b.size()) {
        return false;
    }
    char diff = 0;
    for (int i = 0; i < a.size(); ++i) {
        diff |= static_cast<char>(a[i] ^ b[i]);
    }
    return diff == 0;
}

} // namespace

QString OfflineLoginIntegrity::defaultHmacKey()
{
    return QStringLiteral("1^z4UGNHULekb*_msnrJZUpIJC.H_*.Npw22NJ9Vtv2_Yqro");
}

QString OfflineLoginIntegrity::buildCanonical(const QJsonObject &root)
{
    QString canonical;
    canonical.reserve(512);
    appendLine(canonical, QStringLiteral("app_id"),
               fieldString(root, "app_id", "appId"));
    appendLine(canonical, QStringLiteral("app_name"),
               fieldString(root, "app_name", "appName"));
    appendLine(canonical, QStringLiteral("expire_at"),
               fieldString(root, "expire_at", "expireAt"));
    appendLine(canonical, QStringLiteral("host"), fieldString(root, "host"));
    appendLine(canonical, QStringLiteral("local_password_hash"),
               fieldString(root, "local_password_hash", "localPasswordHash"));
    appendLine(canonical, QStringLiteral("local_user_name"),
               fieldString(root, "local_user_name", "localUserName"));
    appendLine(canonical, QStringLiteral("password"), fieldString(root, "password"));
    appendLine(canonical, QStringLiteral("spa_key"),
               fieldString(root, "spa_key", "spaKey"));
    appendLine(canonical, QStringLiteral("spa_port"),
               fieldInt(root, "spa_port", "spaPort"));
    appendLine(canonical, QStringLiteral("srv_port"),
               fieldInt(root, "srv_port", "srvPort"));
    appendLine(canonical, QStringLiteral("user_name"),
               fieldString(root, "user_name", "userName"));
    return canonical;
}

QString OfflineLoginIntegrity::signCanonical(const QString &canonical, const QString &hmacKey)
{
    const QByteArray mac =
        HmacSha256::sign(hmacKey.toUtf8(), canonical.toUtf8());
    return QString::fromLatin1(mac.toBase64());
}

bool OfflineLoginIntegrity::verifyPayload(const QJsonObject &root, const QString &hmacKey)
{
    const QString payloadHmac =
        fieldString(root, "payload_hmac", "payloadHmac");
    if (payloadHmac.isEmpty() || hmacKey.isEmpty()) {
        return false;
    }
    const QString canonical = buildCanonical(root);
    const QByteArray expected = QByteArray::fromBase64(signCanonical(canonical, hmacKey).toLatin1());
    const QByteArray actual = QByteArray::fromBase64(payloadHmac.toLatin1());
    return !expected.isEmpty() && secureEquals(expected, actual);
}

} // namespace vpn

File: ruoyi-modules/ruoyi-yianlian/src/main/java/com/ruoyi/yianlian/utils/OfflineLoginIntegrity.java
package com.ruoyi.yianlian.utils;

import com.ruoyi.common.core.utils.StringUtils;
import com.ruoyi.yianlian.domain.vo.VpnOfflineLoginDatPayload;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Base64;

/**
 * 离线登录 .dat 完整性签名（HMAC-SHA256）
 */
public final class OfflineLoginIntegrity
{
    private static final String HMAC_ALGORITHM = "HmacSHA256";

    private OfflineLoginIntegrity()
    {
    }

    public static String buildCanonical(VpnOfflineLoginDatPayload payload)
    {
        StringBuilder sb = new StringBuilder(256);
        appendLine(sb, "app_id", payload.getAppId());
        appendLine(sb, "app_name", payload.getAppName());
        appendLine(sb, "expire_at", payload.getExpireAt());
        appendLine(sb, "host", payload.getHost());
        appendLine(sb, "local_password_hash", payload.getLocalPasswordHash());
        appendLine(sb, "local_user_name", payload.getLocalUserName());
        appendLine(sb, "password", payload.getPassword());
        appendLine(sb, "spa_key", payload.getSpaKey());
        appendLine(sb, "spa_port", payload.getSpaPort());
        appendLine(sb, "srv_port", payload.getSrvPort());
        appendLine(sb, "user_name", payload.getUserName());
        return sb.toString();
    }

    public static String sign(String canonical, String hmacKey)
    {
        if (StringUtils.isEmpty(hmacKey))
        {
            throw new IllegalArgumentException("HMAC 密钥未配置");
        }
        try
        {
            Mac mac = Mac.getInstance(HMAC_ALGORITHM);
            mac.init(new SecretKeySpec(hmacKey.getBytes(StandardCharsets.UTF_8), HMAC_ALGORITHM));
            byte[] digest = mac.doFinal(canonical.getBytes(StandardCharsets.UTF_8));
            return Base64.getEncoder().encodeToString(digest);
        }
        catch (Exception e)
        {
            throw new IllegalStateException("离线登录 HMAC 签名失败: " + e.getMessage(), e);
        }
    }

    public static boolean verify(String canonical, String hmacKey, String payloadHmac)
    {
        if (StringUtils.isEmpty(payloadHmac))
        {
            return false;
        }
        String expected = sign(canonical, hmacKey);
        return MessageDigest.isEqual(
            expected.getBytes(StandardCharsets.UTF_8),
            payloadHmac.getBytes(StandardCharsets.UTF_8));
    }

    private static void appendLine(StringBuilder sb, String key, Object value)
    {
        sb.append(key).append('=');
        if (value == null)
        {
            sb.append("");
        }
        else
        {
            sb.append(value);
        }
        sb.append('\n');
    }
}
